dao.go

```diff
@@ -1,53 +1,113 @@
 package main
 
-type Dao struct {
-	data map[int]Device
+import (
+	"context"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"log"
+	"os"
+)
 
-	indexer int
+type Dao struct {
+	mongoClient *mongo.Client
+	collection  *mongo.Collection
+	ctx         context.Context
 }
 
 type DeviceDao interface {
-	AddDevice(device *DevicePayload) (int, error)
-	GetDevice(id int) (*Device, error)
+	AddDevice(device *DevicePayload) (primitive.ObjectID, error)
+	GetDevice(id string) (*Device, error)
 	GetPaginatedDevices(limit int, page int) ([]Device, error)
 	GetAllDevices() ([]Device, error)
 }
 
 func NewDao() *Dao {
-	d := Dao{data: make(map[int]Device)}
-	return &d
+	mongodbURI := os.Getenv("MONGODB_URI")
+	client, err := mongo.NewClient(options.Client().ApplyURI(mongodbURI))
+	if err != nil {
+		log.Panicf("couldnt create client with the uri: %s :%+v", mongodbURI, err.Error())
+	}
+	collection := client.Database(os.Getenv("MONGODB_NAME")).Collection("devices")
+	dao := &Dao{
+		mongoClient: client,
+		collection:  collection,
+		ctx:         context.Background(),
+	}
+	dao.connect()
+	return dao
+}
+
+func (db *Dao) connect() {
+	err := db.mongoClient.Connect(db.ctx)
+	if err != nil {
+		log.Panicf("couldn't connect to db: %+v", err.Error())
+	}
+	err = db.mongoClient.Ping(db.ctx, readpref.Primary())
+	if err != nil {
+		log.Panicf("connection with db was not established properly: %+v", err.Error())
+	}
 }
 
-func (d *Dao) AddDevice(device *DevicePayload) (int, error) {
-	d.indexer++
-	dev := &Device{
-		Id:       d.indexer,
+func (db *Dao) AddDevice(device *DevicePayload) (primitive.ObjectID, error) {
+	dev := Device{
+		Id:       primitive.NewObjectID(),
 		Name:     device.Name,
 		Value:    device.Value,
 		Interval: device.Interval,
 	}
+	result, err := db.collection.InsertOne(db.ctx, dev)
+	if err != nil {
+		log.Printf("%v was not added to db: %+v", dev, err.Error())
+		return [12]byte{}, err
+	}
 
-	d.data[dev.Id] = *dev
-	return dev.Id, nil
+	return result.InsertedID.(primitive.ObjectID), nil
 }
 
-func (d *Dao) GetDevice(id int) (*Device, error) {
-	if device, ok := d.data[id]; ok {
-		return &device, nil
+func (db *Dao) GetDevice(id string) (*Device, error) {
+	objID, err := primitive.ObjectIDFromHex(id)
+	if err != nil {
+		return nil, err
+	}
+	findResult := db.collection.FindOne(db.ctx, bson.M{"_id": objID})
+	if err := findResult.Err(); err != nil {
+		return nil, err
 	}
-	return nil, nil
+	dev := Device{}
```

Comment:
In case device is not found, we should return nil. This statement creates new, empty object - you will always return something
```suggestion
	var dev *Device
```